### PR TITLE
VMO-4131-router-test-functions

### PR DESCRIPTION
**Pull request by seifertk**
Some functions unimplemented since they rely on database-like functionality -- but this is only a few.
They can be implemented on by the consumer of the library by extending or composing the RouterTest handler and injecting into the evaluator.
Tests included for every implemented function in `tests/Evaluator/MethodHandler/RouterTestHandlerTest.php`. You will likely need to click on it in the diff viewer since it's a big diff (1034 lines). Test cases were adapted from RapidPro documentation.

### CODE REVIEW

File: .ci/5.5/composer.json [hunk 1]
```diff
@@ -10,7 +10,9 @@
         "php": "~5.5|~7.1",
         "ext-bcmath": "*",
         "nesbot/carbon": "^1.38|^2.0",
-        "illuminate/support": "~5.1"
+        "illuminate/support": "~5.1",
+        "najidev/permutation": "^1.0",
+        "giggsey/libphonenumber-for-php": "^8.12"
```
Comment (AlteaDown) [inline .ci/5.5/composer.json:15]: Shouldn't these be alphabetically sorted?

File: src/php/Evaluator/MethodNodeEvaluator.php [hunk 2]
```diff
@@ -72,7 +74,20 @@ public function evaluate(Node $node, $context)
...
-        return call_user_func_array([$handler, $call], $args);
+
+        $result = call_user_func_array([$handler, $call], $args);
+
+        if (isset($node['chain'])) {
+            foreach ($node['chain'] as $chain) {
+                if ($result instanceof Chainable) {
+                    $result = $result->chain($chain);
+                } else {
+                    throw new MethodNodeException("Must chain on Chainable object, got " . \gettype($result));
```
Comment (AlteaDown) [inline src/php/Evaluator/MethodNodeEvaluator.php:85]: On PHP 5.6+ I believe you can import gettype with a use statement, so you don't have to refer to it by it's absolute namespace.

File: src/php/Evaluator/MethodNodeEvaluator/Contract/RouterTest.php [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+    public function has_beginning($text, $beginning);
+
+    /**
+     * Tests whether the category of a result on of the passed in categories
+     *
+     * @param array|Traversable|Node $result
+     * @param string ...$categories
+     * @return TestResult
+     */
+    public function has_category($result);
```
Comment (AlteaDown) [inline src/php/Evaluator/MethodNodeEvaluator/Contract/RouterTest.php]: where is the variadic param `string ...$categories`?

File: src/php/Evaluator/MethodNodeEvaluator/Contract/RouterTest.php [hunk 4]
```diff
@@ -0,0 +1,294 @@
+<?php
+
+namespace Viamo\Floip\Evaluator\MethodNodeEvaluator\Contract;
+
+use Viamo\Floip\Evaluator\MethodNodeEvaluator\Contract\EvaluatesMethods;
+use Viamo\Floip\Evaluator\MethodNodeEvaluator\Contract\TestResult;
+
+interface RouterTest extends EvaluatesMethods
```
Comment (AlteaDown) [inline src/php/Evaluator/MethodNodeEvaluator/Contract/RouterTest.php]: Liking the use of PHPDoc/PHPStan comments here! 😄

**Review by AlteaDown — APPROVED** [review]
Pre-approving pending review of missing param.

File: src/php/Evaluator/MethodNodeEvaluator/Contract/RouterTest.php (hunk 3 above)
Comment (seifertk) [reply src/php/Evaluator/MethodNodeEvaluator/Contract/RouterTest.php]: I wish we could have that :cry: 
But some DNs are still running 5.5.9
So we must just _hint_ the variadic variable and use `func_get_args`

File: src/php/Evaluator/MethodNodeEvaluator/RouterTest.php [hunk 5]
```diff
@@ -0,0 +1,452 @@
...
+use Exception;
+use libphonenumber\PhoneNumberFormat;
+use libphonenumber\PhoneNumberUtil;
+use NajiDev\Permutation\PermutationIterator;
+use Viamo\Floip\Evaluator\Exception\MethodNodeException;
+use Viamo\Floip\Evaluator\MethodNodeEvaluator\Contract\RouterTest as RouterTestInterface;
+use Viamo\Floip\Evaluator\MethodNodeEvaluator\TestResult;
+use Viamo\Floip\Evaluator\Node;
+
+// todo: do we need utf aware string comparison?
```
Comment (GrimFunction) [inline src/php/Evaluator/MethodNodeEvaluator/RouterTest.php]: Any todo should have an associated ticket

File: src/php/Evaluator/MethodNodeEvaluator/Contract/RouterTest.php [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+{
+    const PUNCTUATION=',:;!?.-';
```
Comment (GrimFunction) [inline src/php/Evaluator/MethodNodeEvaluator/Contract/RouterTest.php]: Are you purposely omitting dashes, brackets, braces, parentheses, quotes, and ellipsis?

File: src/php/Evaluator/MethodNodeEvaluator/RouterTest.php [hunk 7]
```diff
@@ -0,0 +1,452 @@
...
+        });
+    }
+
+    public function has_date_lt($text, $max) {
+        return $this->has_date_callback($text, $max, function (Carbon $lhs, Carbon $rhs) {
+            return $lhs->lt($rhs);
+        });
+    }
+
+    // todo: implementation?
```
Comment (GrimFunction) [inline src/php/Evaluator/MethodNodeEvaluator/RouterTest.php]: ticket?

File: src/php/Evaluator/MethodNodeEvaluator/RouterTest.php [hunk 8]
```diff
@@ -0,0 +1,452 @@
...
+        $regex = "/[A-Z0-9a-z._%+-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,64}/";
+        $matches = [];
+        if (preg_match($regex, $text, $matches)) {
+            return new TestResult(true, $matches[0]);
+        } else {
+            return new TestResult;
+        }
+    }
+
+    // todo: implementation once exceptions turn into error objects
```
Comment (GrimFunction) [inline src/php/Evaluator/MethodNodeEvaluator/RouterTest.php]: Probably should bundle these all into one ticket

File: src/php/Evaluator/MethodNodeEvaluator/RouterTest.php [hunk 9]
```diff
@@ -0,0 +1,452 @@
...
+    }
+
+    // todo: implementation
+    public function has_top_intent($result, $name, $confidence) {
+        throw new MethodNodeException('has_top_intent not implemented');
+    }
+
+    // todo: implementation
+    public function has_ward($text, $district, $state) {
+        throw new MethodNodeException('has_top_intent not implemented');
```
Comment (GrimFunction) [inline src/php/Evaluator/MethodNodeEvaluator/RouterTest.php]: should be `has_ward_not_implemented`

File: src/php/Evaluator/MethodNodeEvaluator/RouterTest.php [hunk 10]
```diff
@@ -0,0 +1,452 @@
...
+        if ($result) {
+            return new TestResult(true, $matches[0]);
+        } else {
+            return new TestResult;
+        }
+    }
+
+    // todo: implementation? perhaps up to the consumer
+    public function has_state($text) {
+        throw new MethodNodeException('has_district not implemented');
```
Comment (GrimFunction) [inline src/php/Evaluator/MethodNodeEvaluator/RouterTest.php]: Should be 'has_state not implemented'

File: src/php/Evaluator/MethodNodeEvaluator/Contract/RouterTest.php (hunk 3 above)
Comment (AlteaDown) [reply src/php/Evaluator/MethodNodeEvaluator/Contract/RouterTest.php]: uggghh.... 🤮

File: src/php/Evaluator/MethodNodeEvaluator/Contract/RouterTest.php (hunk 4 above)
Comment (markboots) [inline src/php/Evaluator/MethodNodeEvaluator/Contract/RouterTest.php]: Nice work @seifertk !  One small thing to consider on naming: Flow Interop doesn't have the concept of a "router"; this is only in the RapidPro world.  However, we do need to support all of the test expressions that RapidPro might use in routers... They would come into Exit test expressions in the Flow world.  For that reason, the "RouterTest" class naming might not be ideal here, but it's not a big problem.

Comment (seifertk) [reply src/php/Evaluator/MethodNodeEvaluator/Contract/RouterTest.php]: It's very easy to rename at this point: any suggestions?

Comment (markboots) [reply src/php/Evaluator/MethodNodeEvaluator/Contract/RouterTest.php]: Nope; I trust your creativity :)

File: src/php/Evaluator/MethodNodeEvaluator/Contract/RouterTest.php (hunk 6 above)
Comment (seifertk) [reply src/php/Evaluator/MethodNodeEvaluator/Contract/RouterTest.php]: Following existing RapidPro punctuation rules.

**Pull request merged**